File: JavaSDK/src/main/java/net/csibio/aird/parser/ColumnParser.java
package net.csibio.aird.parser;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import net.csibio.aird.bean.ColumnIndex;
import net.csibio.aird.bean.ColumnInfo;
import net.csibio.aird.bean.common.IntPair;
import net.csibio.aird.bean.common.Xic;
import net.csibio.aird.compressor.ByteTrans;
import net.csibio.aird.compressor.intcomp.VarByteWrapper;
import net.csibio.aird.compressor.sortedintcomp.IntegratedVarByteWrapper;
import net.csibio.aird.constant.SuffixConst;
import net.csibio.aird.enums.ResultCodeEnum;
import net.csibio.aird.exception.ScanException;
import net.csibio.aird.util.AirdMathUtil;
import net.csibio.aird.util.AirdScanUtil;

public class ColumnParser {

    /**
     * the aird file
     */
    public File airdFile;

    /**
     * the airdInfo from the index file.
     */
    public ColumnInfo columnInfo;

    /**
     * mz precision
     */
    public double mzPrecision;

    /**
     * intensity precision
     */
    public double intPrecision;

    /**
     * Random Access File reader
     */
    public RandomAccessFile raf;

    public ColumnParser(String indexPath) throws IOException {
        columnInfo = AirdScanUtil.loadColumnInfo(indexPath);
        if (columnInfo == null) {
            throw new ScanException(ResultCodeEnum.AIRD_COLUMN_INDEX_FILE_PARSE_ERROR);
        }
        this.airdFile = new File(AirdScanUtil.getAirdPathByIndexPath(indexPath));
        try {
            raf = new RandomAccessFile(airdFile, "r");
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            throw new ScanException(ResultCodeEnum.AIRD_FILE_PARSE_ERROR);
        }
        //获取mzPrecision
        mzPrecision = columnInfo.getMzPrecision();
        intPrecision = columnInfo.getIntPrecision();
        parseColumnIndex();
    }

    public void parseColumnIndex() throws IOException {
        List<ColumnIndex> indexList = columnInfo.getIndexList();
        for (ColumnIndex columnIndex : indexList) {
            if (columnIndex.getMzs() == null) {
                byte[] mzsByte = readByte(columnIndex.getStartMzListPtr(), columnIndex.getEndMzListPtr());
                int[] mzsAsInt = decodeAsSortedInteger(mzsByte);
                columnIndex.setMzs(mzsAsInt);
            }
            if (columnIndex.getRts() == null) {
                byte[] rtsByte = readByte(columnIndex.getStartRtListPtr(), columnIndex.getEndRtListPtr());
                int[] rtsAsInt = decodeAsSortedInteger(rtsByte);
                columnIndex.setRts(rtsAsInt);
            }
            if (columnIndex.getSpectraIds() == null) {
                byte[] spectraIdBytes = readByte(columnIndex.getStartSpectraIdListPtr(), columnIndex.getEndSpectraIdListPtr());
                int[] spectraIds = decode(spectraIdBytes);
                columnIndex.setSpectraIds(spectraIds);
            }
            if (columnIndex.getIntensities() == null) {
                byte[] intensityBytes = readByte(columnIndex.getStartIntensityListPtr(), columnIndex.getEndIntensityListPtr());
                int[] intensities = decode(intensityBytes);
                columnIndex.setIntensities(intensities);
            }
        }
    }

    public byte[] readByte(long startPtr, long endPtr) throws IOException {
        int delta = (int) (endPtr - startPtr);
        raf.seek(startPtr);
        byte[] result = new byte[delta];
        raf.read(result);
        return result;
    }

    public byte[] readByte(long startPtr, int delta) throws IOException {
        raf.seek(startPtr);
        byte[] result = new byte[delta];
        raf.read(result);
        return result;
    }

    public Xic calcXicByMz(Double mz, Double mzWindow) throws IOException {
        return calcXic(mz - mzWindow, mz + mzWindow, null, null, null);
    }

    public Xic calcXicByWindow(Double mz, Double mzWindow, Double rt, Double rtWindow, Double precursorMz) throws IOException {
        return calcXic(mz - mzWindow, mz + mzWindow, rt - rtWindow, rt + rtWindow, precursorMz);
    }

    public Xic calcXic(Double mzStart, Double mzEnd, Double rtStart, Double rtEnd, Double precursorMz) throws IOException {
        if (columnInfo.getIndexList() == null || columnInfo.getIndexList().size() == 0) {
            return null;
        }
        ColumnIndex index = null;
        if (precursorMz != null) {
            for (ColumnIndex columnIndex : columnInfo.getIndexList()) {
                if (columnIndex.getRange() != null && columnIndex.getRange().getStart() <= precursorMz && columnIndex.getRange().getEnd() > precursorMz) {
                    index = columnIndex;
                }
            }
        } else {
            index = columnInfo.getIndexList().get(0);
        }
        if (index == null) {
            return null;
        }

        int[] mzs = index.getMzs();
        int start = (int) (mzStart * mzPrecision);
        int end = (int) (mzEnd * mzPrecision);
        IntPair leftMzPair = AirdMathUtil.binarySearch(mzs, start);
        int leftMzIndex = leftMzPair.right();
        IntPair rightMzPair = AirdMathUtil.binarySearch(mzs, end);
        int rightMzIndex = rightMzPair.left();

        int leftRtIndex = 0;
        int rightRtIndex = index.getRts().length - 1;
        if (rtStart != null) {
            IntPair leftRtPair = AirdMathUtil.binarySearch(index.getRts(), (int) (rtStart * 1000));
            leftRtIndex = leftRtPair.right();
        }

        if (rtEnd != null) {
            IntPair rightRtPair = AirdMathUtil.binarySearch(index.getRts(), (int) (rtEnd * 1000));
            rightRtIndex = rightRtPair.left();
        }
        int[] spectraIdLengths = index.getSpectraIds();
        int[] intensityLengths = index.getIntensities();
        int anchorIndex = leftMzIndex / 100000;
        long startPtr = index.getAnchors()[anchorIndex];
        for (int i = anchorIndex * 100000; i < leftMzIndex; i++) {
            startPtr += spectraIdLengths[i];
            startPtr += intensityLengths[i];
        }
        List<Map<Integer, Double>> columnMapList = new ArrayList<>();
        TreeMap<Integer, Double> map = new TreeMap<>();  //key为spectraId，value为intensity
        for (int k = leftMzIndex; k <= rightMzIndex; k++) {
            byte[] spectraIdBytes = readByte(startPtr, spectraIdLengths[k]);
            startPtr += spectraIdLengths[k];
            byte[] intensityBytes = readByte(startPtr, intensityLengths[k]);
            startPtr += intensityLengths[k];
            int[] spectraIds = fastDecodeAsSortedInteger(spectraIdBytes);
            int[] ints = fastDecode(intensityBytes);
            for (int t = 0; t < spectraIds.length; t++) {
                int spectraId = spectraIds[t];
                if (spectraId >= leftRtIndex && spectraId <= rightRtIndex) {
                    double intensity = ints[t];
                    if (intensity < 0) {
                        intensity = Math.pow(2, -intensity / 100000d);
                    }
                    //TODO WXX Optimizer
                    map.merge(spectraId, intensity / intPrecision, Double::sum);
                }
            }

            columnMapList.add(map);
        }
        int rtRange = rightRtIndex - leftRtIndex + 1;
        double[] intensities = new double[rtRange];
        double[] rts = new double[rtRange];
        int iteration = 0;
        for (int i = leftRtIndex; i <= rightRtIndex; i++) {
            intensities[iteration] = map.getOrDefault(i, 0d);
            rts[iteration] = index.getRts()[i] / 1000d;
            iteration++;
        }

        return new Xic(rts, intensities);
    }

    public int[] decodeAsSortedInteger(byte[] origin) {
        return new IntegratedVarByteWrapper().decode(ByteTrans.byteToInt(origin));
    }

    public int[] fastDecodeAsSortedInteger(byte[] origin) {
        return new IntegratedVarByteWrapper().decode(ByteTrans.byteToInt(origin));
    }

    public int[] decode(byte[] origin) {
        return new VarByteWrapper().decode(ByteTrans.byteToInt(origin));
    }

    public int[] fastDecode(byte[] origin) {
        return new VarByteWrapper().decode(ByteTrans.byteToInt(origin));
    }
}

File: CSharpSDK/Parser/ColumnParser.cs
using System;
using System.Collections.Generic;
using System.IO;
using AirdSDK.Beans;
using AirdSDK.Beans.Common;
using AirdSDK.Compressor;
using AirdSDK.Enums;
using AirdSDK.Exception;
using AirdSDK.Utils;
using CSharpFastPFOR.Port;

namespace AirdSDK.Parser;

public class ColumnParser
{
    /**
     * the aird file
     */
    public FileInfo airdFile;

    /**
     * the column index file. JSON format,with cjson
     */
    public FileInfo indexFile;

    /**
     * the airdInfo from the index file.
     */
    public ColumnInfo columnInfo;

    /**
     * mz precision
     */
    public double mzPrecision;

    /**
     * intensity precision
     */
    public double intPrecision;

    /**
     * File reader
     */
    public FileStream fs;

    public ColumnParser(string indexPath)
    {
        indexFile = new FileInfo(indexPath);
        columnInfo = AirdScanUtil.loadColumnInfo(indexFile);
        if (columnInfo == null) throw new ScanException(ResultCodeEnum.AIRD_INDEX_FILE_PARSE_ERROR);

        airdFile = new FileInfo(AirdScanUtil.getAirdPathByColumnIndexPath(indexPath));
        fs = File.OpenRead(airdFile.FullName);

        //获取mzPrecision
        mzPrecision = columnInfo.mzPrecision;
        intPrecision = columnInfo.intPrecision;
        parseColumnIndex();
    }


    public void parseColumnIndex()
    {
        List<ColumnIndex> indexList = columnInfo.indexList;
        foreach (var columnIndex in indexList)
        {
            if (columnIndex.mzs == null)
            {
                byte[] mzsByte = readByte(columnIndex.startMzListPtr, columnIndex.endMzListPtr);
                int[] mzsAsInt = decodeAsSortedInteger(mzsByte);
                columnIndex.mzs = mzsAsInt;
            }

            if (columnIndex.rts == null)
            {
                byte[] rtsByte = readByte(columnIndex.startRtListPtr, columnIndex.endRtListPtr);
                int[] rtsAsInt = decodeAsSortedInteger(rtsByte);
                columnIndex.rts = rtsAsInt;
            }

            if (columnIndex.spectraIds == null)
            {
                byte[] spectraIdBytes = readByte(columnIndex.startSpecrtaIdListPtr, columnIndex.endSpecrtaIdListPtr);
                int[] spectraIds = decode(spectraIdBytes);
                columnIndex.spectraIds = spectraIds;
            }

            if (columnIndex.intensities == null)
            {
                byte[] intensityBytes = readByte(columnIndex.startIntensityListPtr, columnIndex.endIntensityListPtr);
                int[] intensities = decode(intensityBytes);
                columnIndex.intensities = intensities;
            }
        }
    }

    public byte[] readByte(long startPtr, long endPtr)
    {
        int delta = (int)(endPtr - startPtr);
        fs.Seek(startPtr, SeekOrigin.Begin);
        byte[] result = new byte[delta];
        fs.Read(result, 0, delta);
        return result;
    }

    public byte[] readByte(long startPtr, int delta)
    {
        fs.Seek(startPtr, SeekOrigin.Begin);
        byte[] result = new byte[delta];
        fs.Read(result, 0, delta);
        return result;
    }

    public Xic calcXicByMz(double mz, double mzWindow)
    {
        return calcXic(mz - mzWindow, mz + mzWindow, null, null, null);
    }

    public Xic calcXicByWindow(double mz, double mzWindow, double? rt, double? rtWindow, double? precursorMz)
    {
        double? rtStart = null;
        double? rtEnd = null;
        if (rt != null && rtWindow != null)
        {
            rtStart = rt - rtWindow;
            rtEnd = rt + rtWindow;
        }
        return calcXic(mz - mzWindow, mz + mzWindow, rtStart, rtEnd, precursorMz);
    }

    public Xic calcXic(double mzStart, double mzEnd, double? rtStart, double? rtEnd, double? precursorMz)
    {
        if (columnInfo.indexList == null || columnInfo.indexList.Count == 0)
        {
            return null;
        }

        ColumnIndex index = null;
        if (precursorMz != null)
        {
            foreach (var columnIndex in columnInfo.indexList)
            {
                if (columnIndex.range != null && columnIndex.range.start <= precursorMz &&
                    columnIndex.range.end > precursorMz)
                {
                    index = columnIndex;
                }
            }
        }
        else
        {
            index = columnInfo.indexList[0];
        }

        if (index == null)
        {
            return null;
        }

        int[] mzs = index.mzs;
        int start = (int)(mzStart * mzPrecision);
        int end = (int)(mzEnd * mzPrecision);
        IntPair leftMzPair = AirdMathUtil.binarySearch(mzs, start);
        int leftMzIndex = leftMzPair.right();
        IntPair rightMzPair = AirdMathUtil.binarySearch(mzs, end);
        int rightMzIndex = rightMzPair.left();

        int leftRtIndex = 0;
        int rightRtIndex = index.rts.Length - 1;
        if (rtStart != null)
        {
            IntPair leftRtPair = AirdMathUtil.binarySearch(index.rts, (int)(rtStart * 1000));
            leftRtIndex = leftRtPair.right();
        }

        if (rtEnd != null)
        {
            IntPair rightRtPair = AirdMathUtil.binarySearch(index.rts, (int)(rtEnd * 1000));
            rightRtIndex = rightRtPair.left();
        }

        int[] spectraIdLengths = index.spectraIds;
        int[] intensityLengths = index.intensities;
        long startPtr = index.startPtr;
        for (int i = 0; i < leftMzIndex; i++)
        {
            startPtr += spectraIdLengths[i];
            startPtr += intensityLengths[i];
        }

        List<Dictionary<int, double>> columnMapList = new List<Dictionary<int, double>>();

        for (int k = leftMzIndex; k <= rightMzIndex; k++)
        {
            byte[] spectraIdBytes = readByte(startPtr, spectraIdLengths[k]);
            startPtr += spectraIdLengths[k];
            byte[] intensityBytes = readByte(startPtr, intensityLengths[k]);
            startPtr += intensityLengths[k];
            int[] spectraIds = decodeAsSortedInteger(spectraIdBytes);
            int[] ints = decode(intensityBytes);
            Dictionary<int, double> map = new Dictionary<int, double>();
            //解码intensity
            for (int t = 0; t < spectraIds.Length; t++)
            {
                if (spectraIds[t] >= leftRtIndex && spectraIds[t] <= rightRtIndex)
                {
                    double intensity = ints[t];
                    if (intensity < 0)
                    {
                        intensity = Math.Pow(2, -intensity / 100000d);
                    }

                    map[spectraIds[t]] = intensity / intPrecision;
                }
            }

            columnMapList.Add(map);
        }

        int rtRange = rightRtIndex - leftRtIndex + 1;
        double[] intensities = new double[rtRange];
        double[] rts = new double[rtRange];
        int iteration = 0;
        for (int i = leftRtIndex; i <= rightRtIndex; i++)
        {
            double intensity = 0;
            for (int j = 0; j < columnMapList.Count; j++)
            {
                if (columnMapList[j].ContainsKey(i))
                {
                    intensity += columnMapList[j][i];
                }
            }

            intensities[iteration] = intensity;
            rts[iteration] = index.rts[i] / 1000d;
            iteration++;
        }
        
        return new Xic(rts, intensities);
    }

    public int[] decodeAsSortedInteger(byte[] origin)
    {
        if (origin.Length > 16)
        {
            return new IntegratedVarByteWrapper().decode(ByteTrans.byteToInt(new ZstdWrapper().decode(origin)));
        }
        else
        {
            return ByteTrans.byteToInt(origin);
        }
    }

    public int[] decode(byte[] origin)
    {
        if (origin.Length > 16)
        {
            return new VarByteWrapper().decode(ByteTrans.byteToInt(new ZstdWrapper().decode(origin)));
        }
        else
        {
            return ByteTrans.byteToInt(origin);
        }
    }
}